### PR TITLE
Don't be strict with enums.

### PR DESCRIPTION
**Pull request by eduardosm**
X11 enum types are no longer converted to Rust enums. Instead, they are converted to newtypes that wrap an integer type and an associated constant is defined for each variant.

This would fix https://github.com/psychon/x11rb/issues/544 and https://github.com/psychon/x11rb/issues/458.

Enums with repeated values are now treated the same, so it would also fix https://github.com/psychon/x11rb/issues/384.

@psychon What do you think about this solution?

### CODE REVIEW

**Comment by psychon** [comment]
> @psychon What do you think about this solution?

That moves the problem to the library user... and... I'm not quite sure what happens when one wants to match on an enum. That becomes impossible, doesn't it? (Also: First impression is that I can easily make some of those new `unwrap()`s in request sending panic)

So far, we only found two issues, I think:
- XInput sends a value 0 for a field which may not have value 0
- Some GetProperty something sends an invalid reply for non-existing properties (was that also in XInput?)

I'd rather fix those two cases in the XML than simply stopping to validate the garbage that the X11 server sends and moving the problem to the library user.

What do you think?
(But apprently we do not have such a `match` example in our examples? That's surprising...)

**Comment by eduardosm** [comment]
> That moves the problem to the library user... and... I'm not quite sure what happens when one wants to match on an enum. That becomes impossible, doesn't it?

You can still match as before, associated constants can be used as patterns in match arms. Currently enums are `non_exhaustive`, so library users already need to handle the `_` case.

> Some GetProperty something sends an invalid reply for non-existing properties (was that also in XInput?)

This is https://github.com/psychon/x11rb/pull/505, which I forgot to mention. This still does not fix the case for switches, which will fail if the discriminant has an unknown value. Fixing that would require adding an extra `Invalid` variant to case switches (bitcase switches are unaffected).

> I'd rather fix those two cases in the XML than simply stopping to validate the garbage that the X11 server sends and moving the problem to the library user.

My main concern is that I don't trust server implementations to be always strict with the definitions. It wouldn't surprise me if there are other not-yet-found situations where the server sends invalid values, specially in situation when the enum value is not relevant.

Regarding handling garbage:
* If the server is just sending random data, it will be caught sooner or later by other means.
* If it is an isolated bug in the server, the application may want to try to workaround it instead of quitting (as would happen if the error is propagated with the `?` operator), which I guess that it is what libxcb-based applications do.

**Comment by psychon** [comment]
> the application may want to try to workaround it instead of quitting

Random thought: Transition the examples to get raw events and then parse them in a second step. That way, parsing errors can be told apart from connection errors without `match`ing on the errors. Then, the parsing errors can easily be made non-fatal (which really sounds like a good idea, you are right).

I looked backed and think I found the issue that made us generate all this enum-parsing and I think I found a good point:

> [Having enums as e.g. `u16`] makes working with those structures more difficult in some cases, e.g. when dumping their contents.

https://github.com/psychon/x11rb/issues/253#issuecomment-596137196

So.... hrm. That means we need to
- either stop `#[derive(Debug)]`ing and instead generate an implementation that pretty-prints
- or do something like `enum Foo { A, B, C, Unknown(u16) }`.

This last idea is a bad one, so a generated `Debug` implementation it is, I guess.

Anyway, I guess all of this is future work after this change was merged. I'll take another look at the diff and ponder about things.
(It's sad that the enum variants have different names than the constants and we end up with a pointless API break, but such are Rust's rules...)

File: src/protocol/damage.rs
Comment (psychon) [inline src/protocol/damage.rs:49]: What would this impl be useful for? I feel like it should just be dropped.

File: src/protocol/damage.rs
Comment (psychon) [inline src/protocol/damage.rs:95]: This new impl also seems useless.

File: src/rust_connection/xauth.rs
Comment (psychon) [inline src/rust_connection/xauth.rs:25]: Can/should this `Family` enum be replaced with a newtype around `u16`, too? It already has an `Unknown(u16)` variant, so I guess it can just be turned into such.

**Review by psychon — CHANGES_REQUESTED** [review]

File: src/errors.rs
Comment (eduardosm) [reply src/errors.rs:76]: I thought we needed it, but we don't, so I'll make it disappear again :)

File: src/protocol/xproto.rs
Comment (eduardosm) [reply src/protocol/xproto.rs:5792]: I guess this is `From<Gravity> for Option<bool>`, so it returns `None` when the value is `> 1`. I guess all this `bool` conversion thing can be cleaned up.

File: src/protocol/damage.rs
Comment (eduardosm) [reply src/protocol/damage.rs:49]: As I said above, I'll try to avoid useless `bool` conversions.

File: src/protocol/xproto.rs
Comment (psychon) [reply src/protocol/xproto.rs]: Do we even need `From<something too small to fit all variants> for Enum`? When would such an impl be useful?

File: src/protocol/damage.rs
Comment (psychon) [inline src/protocol/damage.rs:87]: I can't comment on something a few lines outside, so I leave this comment here:
```rust
impl TryFrom<u16> for ReportLevel {
    type Error = ParseError;
    fn try_from(value: u16) -> Result<Self, Self::Error> {
        u8::try_from(value).or(Err(ParseError::InvalidValue)).map(Self)
    }
}
```
Now that we do not really check the value any more, is `ConversionFailed` more fitting than `InvalidValue`? The docs for the two are:
```
    /// A value did not fit.
    ///
    /// This error can e.g. happen when a value that was received from the X11 server does not fit
    /// into an `usize`.
    ConversionFailed,

    /// A value was outside of its valid range.
    ///
    /// When parsing the value of an enumeration, not all possible integer values have a defined
    /// meaning. This error occurs when an invalid value is encountered in this context.
    ///
    /// For example, `xproto` has an enumeration `Place` with possible values `OnTop` (0) and
    /// `OnBottom` (1). Any value other than 0 or 1 generates an `InvalidValue` when parsing.
    InvalidValue,
```
(Slightly unrelated, but I think the docs for `InvalidValue` need an overhaul. I am sure that there are still places that use it (`<switch>` parsing?), but the example in the docs is now wrong)

File: examples/tutorial.rs
Comment (eduardosm) [reply examples/tutorial.rs:1253]: Unlike enum variants, associated constants can't be imported with `use`, so we have to use the "full path".

File: generator/src/generator/namespace.rs
Comment (eduardosm) [reply generator/src/generator/namespace.rs:4568]: This is to handle cases where the enum has values larger than the wire type. I'm not sure if it is needed. There's https://github.com/psychon/x11rb/issues/543, but it is about a mask enum, which is not handled here.

Comment (eduardosm) [reply generator/src/generator/namespace.rs:4568]: Hm, needed for example here:

`DeviceType` is serialized as `CARD8`:
https://github.com/psychon/x11rb/blob/a29b84d5c9d4db5dd1a08a7343764f6dc3b2c9c9/xcb-proto-1.14-1-g2b3559c/src/xinput.xml#L2379-L2386
`DeviceType` is serialized as `CARD16`:
https://github.com/psychon/x11rb/blob/a29b84d5c9d4db5dd1a08a7343764f6dc3b2c9c9/xcb-proto-1.14-1-g2b3559c/src/xinput.xml#L1752-L1767

So the `DeviceType` newtype wraps an `u16`, even though the defined values for the enum fits an `u8`.

Comment (psychon) [reply generator/src/generator/namespace.rs:4568]: How is that possible? IIRC, we didn't handle that before and an enum could only be converted into `u8` if all its variants fit it into an `u8`.

Comment (eduardosm) [reply generator/src/generator/namespace.rs:4568]: Note that the unwrap should be optimized away in cases where it is not needed.

Comment (psychon) [reply generator/src/generator/namespace.rs:4568]: Well, `DeviceType` has values from 1 to 5. That fits into an `u8`, so the enum implements `From<Self> for u8`. But it also implements this conversion for larger types, i.e. `From<Self> for u16`. Thus, you can just do `u16::from(value)` instead of `Option<u16>::from(value).unwrap()`, or am I missing something?

Comment (eduardosm) [reply generator/src/generator/namespace.rs:4568]: The issue here is that the new type does not wrap an `u8`, it wraps an `u16` (so there isn't `From<Self> for u8`), so it can infallibly parse the case where the wire type is `CARD16`. This is far from ideal, but X11 is hard. I'll try to think a better way of handling this...

Comment (psychon) [reply generator/src/generator/namespace.rs:4568]: Uhm. Are you saying that the problem is turning an `CARD16` into an `u16`? Huh? Or do you mean `CARD8` as the wire type with `u16` as the smallest size that fits the enum? I'm confused and your example doesn't help me much here (since that enum does fit into `u8`).

Comment (eduardosm) [reply generator/src/generator/namespace.rs:4568]: Indeed, the enum fits into an `u8`. However, there are some places where the wire type is actually larger (`u16`), so, our newtype uses an `u16` in order to be able to keep all values that come from the wire. Unfortunately, there are other places where the wire type is `u8`, so converting the newtype to `u8` is fallible.

Comment (psychon) [reply generator/src/generator/namespace.rs:4568]: Ahhhh. Now I understand the problem. Sorry for being so blind. I totally missed that part of the problem.

Also: Urgh.

So, enums in xcb-proto do not have an inherent size / bit length. Perhaps we should go that way and make enums generic? I doubt that will result in nice code, but that's the first idea that came to my mind. (And the list of trait bounds will be huge... at least `T: From<u16>` expresses that something can store at least 16 bits.)

Just a quick first thought. All of this is a mess...

Comment (eduardosm) [reply generator/src/generator/namespace.rs:4568]: I made some quick code to get which enums have this issue:
```
xproto BackingStore
xproto WindowClass
xproto Gravity
xproto StackMode
xproto AutoRepeatMode
render SubPixel
xinput DeviceType
```

Comment (eduardosm) [reply generator/src/generator/namespace.rs]: Most of `unwrap`s are now gone, the remaining only appear for the enums listed above.

File: src/properties.rs
Comment (eduardosm) [reply src/properties.rs:355]: `Gravity` is one of the tricky enums, so let's figure how to handle them before changing this.

File: generator/src/generator/namespace.rs
Comment (psychon) [reply generator/src/generator/namespace.rs]: If it's just six special cases... how about adding them as special cases?

So... render's `SubPixel` appears only once in RENDER:
- in the reply to `QueryPictFormats` as `CARD32`
All other uses are in RandR:
- reply to `SetScreenConfig` as `CARD16`
- reply to `GetOutputInfo` as `CARD8`.
- event `ScreenChangeNotify` as `CARD16`
- struct `OutputChange` as `CARD8`

Okay...
- Your current version is to try to work around this. If I understood correctly, you currently make `SubPixel` a newtype around `u32`. This should actually work fine for `SubPixel` since this enum is only ever read from the wire.
- We could make only the six enums that are affected generic on the underlying type
- We could duplicate them and have `SubPixel`, `SubPixel_u16` and `SubPixel_u32` (I'm bad at naming things)
- [Below I came up with another idea: truncating the value when serialising]

The problem with your current version is when we have to serialise stuff. E.g. `DeviceType` appears in `struct`s as `u8` and `u16` and we have to "do something" there. Hm.

Another random idea would be: Just have a wrapping conversion? The newtype `DeviceType` is around an `u16` and does not implement `Serialize`. Instead, the code could just truncate the `u16` to an `u8`: `value as u8`. This would silently  discard the high bits, but so what? Garbage in, garbage out. Or something like that.

Oorrrr... `ConnectionError` has a `ParseError` variant. So, instead of panicing, sending the request could fail with `ConnectionError::ParseError(ParseError::ConversionFailed)`?

So much for my random ramblings. Perhaps one of these is a good idea. I don't know.
Also, I have a headache today. Sorry.

Comment (eduardosm) [reply generator/src/generator/namespace.rs]: I don't dislike the truncating conversion. It's the simplest to implement and produces the simplest API. The X11 protocol (and implementations) has been accumulating dust for too long and probably it's too late to build a perfect API on top of it.

**Review by psychon — APPROVED** [review]
I guess we need something like this.

I didn't look through all (well, most) of this again, but since we need this and I already went through this before, I guess we should just merge this and then (if needed) clean up afterwards.

I *think* that if you un-draft this PR, mergify should automatically merge this based on this approval. Dunno if that actually works, but you can also just press merge if you want.

I don't know yet when I will have time to look at this more carefully (and also go through the comments that I left previously and check which of them are still open; IIRC some of them were "future work").

**Comment by psychon** [comment]
To get a feeling for what this change means, I ported druid over (CC @jneem). No "special findings", just FYI.
<details><summary>The diff</summary>

```diff
diff --git a/druid-shell/Cargo.toml b/druid-shell/Cargo.toml
index 165221a..cafef94 100644
--- a/druid-shell/Cargo.toml
+++ b/druid-shell/Cargo.toml
@@ -84,7 +84,7 @@ glib = { version = "0.10.1", optional = true }
 glib-sys = { version = "0.10.0", optional = true }
 gtk-sys = { version = "0.10.0", optional = true }
 nix = { version = "0.18.0", optional = true }
-x11rb = { version = "0.6.0", features = ["allow-unsafe-code", "present", "randr", "xfixes"], optional = true }
+x11rb = { version = "0.6.0", features = ["allow-unsafe-code", "present", "randr", "xfixes"], optional = true, git = "https://github.com/psychon/x11rb.git", branch = "not-strict-enums" }
 
 [target.'cfg(target_arch="wasm32")'.dependencies]
 wasm-bindgen = "0.2.67"
diff --git a/druid-shell/src/platform/x11/application.rs b/druid-shell/src/platform/x11/application.rs
index 01ed337..ebdff8a 100644
--- a/druid-shell/src/platform/x11/application.rs
+++ b/druid-shell/src/platform/x11/application.rs
@@ -207,11 +207,11 @@ impl Application {
             // Border width
             0,
             // Window class type
-            WindowClass::InputOnly,
+            WindowClass::INPUT_ONLY,
             // Visual ID
             x11rb::COPY_FROM_PARENT,
             // Window properties mask
-            &CreateWindowAux::new().event_mask(EventMask::StructureNotify),
+            &CreateWindowAux::new().event_mask(EventMask::STRUCTURE_NOTIFY),
         )?
         .check()
         .context("create input-only window")?;
diff --git a/druid-shell/src/platform/x11/error.rs b/druid-shell/src/platform/x11/error.rs
index ec268b4..ef23cdb 100644
--- a/druid-shell/src/platform/x11/error.rs
+++ b/druid-shell/src/platform/x11/error.rs
@@ -31,8 +31,8 @@ impl fmt::Display for Error {
 
 impl std::error::Error for Error {}
 
-impl From<x11rb::protocol::Error> for Error {
-    fn from(err: x11rb::protocol::Error) -> Error {
+impl From<x11rb::x11_utils::X11Error> for Error {
+    fn from(err: x11rb::x11_utils::X11Error) -> Error {
         Error::XError(Arc::new(x11rb::errors::ReplyError::X11Error(err)))
     }
 }
diff --git a/druid-shell/src/platform/x11/util.rs b/druid-shell/src/platform/x11/util.rs
index 98729ba..9c2714e 100644
--- a/druid-shell/src/platform/x11/util.rs
+++ b/druid-shell/src/platform/x11/util.rs
@@ -42,10 +42,10 @@ pub fn refresh_rate(conn: &Rc<XCBConnection>, window_id: Window) -> Option<f64>
                 let flags = mode_info.mode_flags;
                 let vtotal = {
                     let mut val = mode_info.vtotal;
-                    if (flags & u32::from(ModeFlag::DoubleScan)) != 0 {
+                    if (flags & u32::from(ModeFlag::DOUBLE_SCAN)) != 0 {
                         val *= 2;
                     }
-                    if (flags & u32::from(ModeFlag::Interlace)) != 0 {
+                    if (flags & u32::from(ModeFlag::INTERLACE)) != 0 {
                         val /= 2;
                     }
                     val
diff --git a/druid-shell/src/platform/x11/window.rs b/druid-shell/src/platform/x11/window.rs
index 5ebf1d7..788f651 100644
--- a/druid-shell/src/platform/x11/window.rs
+++ b/druid-shell/src/platform/x11/window.rs
@@ -159,7 +159,7 @@ impl WindowBuilder {
         // Replace the window's WM_PROTOCOLS with the following.
         let protocols = [atoms.WM_DELETE_WINDOW];
         conn.change_property32(
-            PropMode::Replace,
+            PropMode::REPLACE,
             window_id,
             atoms.WM_PROTOCOLS,
             AtomEnum::ATOM,
@@ -215,13 +215,13 @@ impl WindowBuilder {
         let visual_id = visual_type.visual_id;
 
         let cw_values = xproto::CreateWindowAux::new().event_mask(
-            EventMask::Exposure
-                | EventMask::StructureNotify
-                | EventMask::KeyPress
-                | EventMask::KeyRelease
-                | EventMask::ButtonPress
-                | EventMask::ButtonRelease
-                | EventMask::PointerMotion,
+            EventMask::EXPOSURE
+                | EventMask::STRUCTURE_NOTIFY
+                | EventMask::KEY_PRESS
+                | EventMask::KEY_RELEASE
+                | EventMask::BUTTON_PRESS
+                | EventMask::BUTTON_RELEASE
+                | EventMask::POINTER_MOTION,
         );
 
         // Create the actual window
@@ -247,7 +247,7 @@ impl WindowBuilder {
             // Border width
             0,
             // Window class type
-            WindowClass::InputOutput,
+            WindowClass::INPUT_OUTPUT,
             // Visual ID
             visual_id,
             // Window properties mask
@@ -296,7 +296,7 @@ impl WindowBuilder {
         let pid = nix::unistd::Pid::this().as_raw();
         if let Ok(pid) = u32::try_from(pid) {
             conn.change_property32(
-                xproto::PropMode::Replace,
+                xproto::PropMode::REPLACE,
                 id,
                 atoms._NET_WM_PID,
                 AtomEnum::CARDINAL,
@@ -337,8 +337,8 @@ impl WindowBuilder {
             // We use the CompleteNotify events to schedule the next frame, and the IdleNotify
             // events to manage our buffers.
             let id = conn.generate_id()?;
-            use x11rb::protocol::present::EventMask::*;
-            conn.present_select_input(id, window_id, CompleteNotify | IdleNotify)?
+            use x11rb::protocol::present::EventMask;
+            conn.present_select_input(id, window_id, EventMask::COMPLETE_NOTIFY | EventMask::IDLE_NOTIFY)?
                 .check()
                 .context("set present event mask")?;
 
@@ -745,12 +745,12 @@ impl Window {
         let conn = self.app.connection();
         log_x11!(conn.configure_window(
             self.id,
-            &xproto::ConfigureWindowAux::new().stack_mode(xproto::StackMode::Above),
+            &xproto::ConfigureWindowAux::new().stack_mode(xproto::StackMode::ABOVE),
         ));
         log_x11!(conn.set_input_focus(
-            xproto::InputFocus::PointerRoot,
+            xproto::InputFocus::POINTER_ROOT,
             self.id,
-            xproto::Time::CurrentTime,
+            xproto::Time::CURRENT_TIME,
         ));
     }
 
@@ -815,14 +815,14 @@ impl Window {
         // what it really is. WM_LOCALE_NAME might be involved. Hopefully, nothing cares about this
         // as long as _NET_WM_NAME is also set (which uses UTF8).
         log_x11!(self.app.connection().change_property8(
-            xproto::PropMode::Replace,
+            xproto::PropMode::REPLACE,
             self.id,
             AtomEnum::WM_NAME,
             AtomEnum::STRING,
             title.as_bytes(),
         ));
         log_x11!(self.app.connection().change_property8(
-            xproto::PropMode::Replace,
+            xproto::PropMode::REPLACE,
             self.id,
             self.atoms._NET_WM_NAME,
             self.atoms.UTF8_STRING,
@@ -1220,7 +1220,7 @@ impl PresentData {
             // idle fence
             x11rb::NONE,
             // present options
-            x11rb::protocol::present::Option::None.into(),
+            x11rb::protocol::present::Option::NONE.into(),
             // target msc (0 means present at the next time that msc % divisor == remainder)
             0,
             // divisor
@@ -1265,7 +1265,7 @@ fn mouse_buttons(mods: u16) -> MouseButtons {
         // BUTTON_MASK_4/5 do not work: they are for scroll events.
     ];
     for (mask, button) in button_masks {
-        if mods & (*mask as u16) != 0 {
+        if mods & u16::from(*mask) != 0 {
             buttons.insert(*button);
         }
     }
@@ -1277,18 +1277,18 @@ fn mouse_buttons(mods: u16) -> MouseButtons {
 fn key_mods(mods: u16) -> Modifiers {
     let mut ret = Modifiers::default();
     let mut key_masks = [
-        (xproto::ModMask::Shift, Modifiers::SHIFT),
-        (xproto::ModMask::Control, Modifiers::CONTROL),
+        (xproto::ModMask::SHIFT, Modifiers::SHIFT),
+        (xproto::ModMask::CONTROL, Modifiers::CONTROL),
         // X11's mod keys are configurable, but this seems
         // like a reasonable default for US keyboards, at least,
         // where the "windows" key seems to be MOD_MASK_4.
         (xproto::ModMask::M1, Modifiers::ALT),
         (xproto::ModMask::M2, Modifiers::NUM_LOCK),
         (xproto::ModMask::M4, Modifiers::META),
-        (xproto::ModMask::Lock, Modifiers::CAPS_LOCK),
+        (xproto::ModMask::LOCK, Modifiers::CAPS_LOCK),
     ];
     for (mask, modifiers) in &mut key_masks {
-        if mods & (*mask as u16) != 0 {
+        if mods & u16::from(*mask) != 0 {
             ret |= *modifiers;
         }
     }

```

</details

**Comment by eduardosm** [comment]
I don't have much time neither, so I'll go through the remaining comments later.

**Comment by andrewshadura** [comment]
I think this broke e.g. `xinput_xi_query_device` to which I should be able to pass `Device::ALL` but I cannot.

**Comment by andrewshadura** [comment]
I also can no longer create a `HierarchyMask` from its integer representation because its field is private.

**Comment by psychon** [comment]
> I think this broke e.g. xinput_xi_query_device to which I should be able to pass Device::ALL but I cannot.

Thanks for testing & thanks for the report. #554 should deal with that.

> I also can no longer create a HierarchyMask from its integer representation because its field is private and there’s no appropriate From.

I think you are looking for this impl which does exist: `impl From<u8> for HierarchyMask`
If this does not work for you, please tell me so. I am not sure if we got this right, because the `HierarchyMask` enum is used twice in the XML and both times it is used as the mask for a field with type `CARD32` and not `CARD8`, so something might be wrong here, but I am not sure. I bet you know xinput better than I do and can report back.

**Pull request merged**
